File: src/vct/col/ast/ParallelAtomic.java
package vct.col.ast;

import java.util.ArrayList;

public class ParallelAtomic extends ASTNode {

  @Override
  public <R,A> R accept_simple(ASTMapping1<R,A> map,A arg){
    return map.map(this,arg);
  }

  
  @Override
  public <T> void accept_simple(ASTVisitor<T> visitor){
    try {
      visitor.visit(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }
  
  @Override
  public <T> T accept_simple(ASTMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
    }
      throw t;
    }
  }
 
 
  public final BlockStatement block; 
  
  public final ArrayList<ASTNode> sync_list;
  
  
  public ParallelAtomic(BlockStatement block,ASTNode ... sync){
    sync_list=new ArrayList();
    for(ASTNode s:sync){
      sync_list.add(s);
    }
    this.block=block;
  }

}

File: vercors/src/main/java/vct/col/ast/ParallelAtomic.scala
package vct.col.ast

import vct.col.util.VisitorHelper

case class ParallelAtomic (val block:BlockStatement, val synclist:List[ASTNode]) extends ASTNode with VisitorHelper {
  def this(block:BlockStatement, synclist:Array[ASTNode]) = this(block, synclist.toList)
  
  /** Yields a copy (as an array) of the sync list (for Java interoperability) */
  def synclistAsArray = synclist.toArray
  
  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))
}
